crates/runtime/src/http.rs

```diff
@@ -90,33 +93,64 @@ where
 
     runtime_metrics::spiced_runtime::HTTP_SERVER_START.add(1, &[]);
 
-    loop {
-        let stream = match listener.accept().await {
-            Ok((stream, _)) => stream,
-            Err(e) => {
-                tracing::debug!("Error accepting connection to serve HTTP request: {e}");
-                continue;
-            }
-        };
+    let shutdown_signal = shutdown_signal.unwrap_or_default();
 
-        match tls_config {
-            Some(ref config) => {
-                let acceptor = TlsAcceptor::from(Arc::clone(&config.server_config));
-                process_tls_tcp_stream(stream, acceptor, routes.clone());
-            }
-            None => {
-                process_tcp_stream(stream, routes.clone());
+    let (shutdown_notify, _) = watch::channel(());
+
+    loop {
+        tokio::select! {
+            conn = listener.accept() => {
+                let stream = match conn {
+                    Ok((stream, _)) => stream,
+                    Err(e) => {
+                        tracing::debug!("Error accepting connection to serve HTTP request: {e}");
+                        continue;
+                    }
+                };
+
+                match tls_config {
+                    Some(ref config) => {
+                        let acceptor = TlsAcceptor::from(Arc::clone(&config.server_config));
+                        process_tls_tcp_stream(stream, acceptor, routes.clone(), shutdown_notify.subscribe());
+                    }
+                    None => {
+                        process_tcp_stream(stream, routes.clone(), shutdown_notify.subscribe());
+                    }
+                };
+            },
+            () = shutdown_signal.cancelled() => {
+                tracing::debug!("Received shutdown signal");
+                drop(listener); // stop accepting new connections while shutting down
+                let num_active = shutdown_notify.receiver_count();
+                if num_active > 0 {
+                    tracing::info!(
+                        "Detected {num_active} active requests. Waiting for completion before shutting down..."
+                    );
+                }
+                shutdown_notify.send(()).ok();
+                // Wait for all active connections to close
+                shutdown_notify.closed().await;
+                break;
             }
-        };
+        }
     }
+
+    tracing::debug!("Stopped");
+
+    Ok(())
 }
 
-fn process_tls_tcp_stream(stream: TcpStream, acceptor: TlsAcceptor, routes: Router) {
+fn process_tls_tcp_stream(
```

Comment:
Is there anyway to test this?

Reply:
@phillipleblanc - added unit tests: https://github.com/spiceai/spiceai/pull/5102/commits/465ca6d67268b80c6eedcb59290f7b8ce2ce92f5